### PR TITLE
AgentJanitor Rework

### PR DESCRIPTION
**Pull request by tylerwowen**
## Problems
1. AgentJanitor cleans up based on the agents_and_hosts table, but it's not complete. There are hosts that' never had an agent. So it should look at the hosts table as well. 
2. The order of processing was wrong. Stale hosts should be a subset of unreachable hosts. The current implementation processes unreachable hosts first then stale hosts. This can result in stale hosts being processed twice. 

## Improvements
Now the `AgentJanitor` does 3 things:
1. `processStaleHosts();`
2. `determineStaleHostCandidates();`
3. `cleanUpAgentlessHosts();`

The first 2 remain the same, except the order has been swapped. This resolves problem 2. The third task solves problem 1.

###  cleanUpAgentlessHosts
This method queries the DB via the new API `HostDAO:getStaleAgentlessHostIds` to get a list of host IDs without associated agents. The list is filtered by `last_update` so only stale hosts will be process. 

```SQL
SELECT DISTINCT
    hosts.host_id
FROM
    hosts
    LEFT JOIN hosts_and_agents ON hosts.host_id = hosts_and_agents.host_id
WHERE
    hosts.last_update < ?
    AND hosts_and_agents.host_id IS NULL
ORDER BY
    hosts.last_update DESC
LIMIT
    ?
```

## Tests and validations
Deployed to dev1 and manually validated the status by reviewing corresponding logs. 

CDP-6636

### CODE REVIEW

File: deploy-service/common/src/main/java/com/pinterest/deployservice/handler/HostHandler.java [hunk 1]
```diff
@@ -27,15 +24,36 @@ public HostHandler(ServiceContext serviceContext) {
...
             hostDAO.deleteAllById(hostId);
+        } catch (Exception e) {
+            hasException = true;
+            LOG.error("Failed to remote host record " + hostId, e);
+        }
+        try {
             agentDAO.deleteAllById(hostId);
+        } catch (Exception e) {
+            hasException = true;
+            LOG.error("Failed to remote host record " + hostId, e);
```
[the remark this answers is not in the document]
Comment (tylerwowen) [reply deploy-service/common/src/main/java/com/pinterest/deployservice/handler/HostHandler.java:39]: SG

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/ConfigHelper.java [hunk 2]
```diff
@@ -91,7 +91,7 @@ public class ConfigHelper {
     private static final int DEFAULT_PERIOD = 30;
     private static final int DEFAULT_MAX_STALE_HOST_THRESHOLD = 600; // 10 mins
     private static final int DEFAULT_MIN_STALE_HOST_THRESHOLD = 150;
-    private static final int DEFAULT_LAUNCH_LATENCY_THRESHOLD = 600;
+    private static final int DEFAULT_LAUNCH_LATENCY_THRESHOLD = 2 * 3600; // 2 hours
```
[the remark this answers is not in the document]
Comment (tylerwowen) [reply deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/ConfigHelper.java:94]: Good catch, it's unintentional

[the remark this answers is not in the document]
Comment (tylerwowen) [reply deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/ConfigHelper.java]: reverted. changed in `AgentJanitor` instead

File: deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java [hunk 3]
```diff
@@ -15,119 +15,198 @@
...
+
+        Set<String> terminatedHosts = getTerminatedHostsFromSource(new ArrayList<>(staleHostMap.keySet()));
+        for (String staleId : staleHostMap.keySet()) {
             if (terminatedHosts.contains(staleId)) {
                 removeStaleHost(staleId);
             } else {
-                markUnreachableHost(staleId);
+                HostAgentBean hostAgent = staleHostMap.get(staleId);
+                if (isHostStale(hostAgent)) {
+                    LOG.warn("Agent ({}) is stale (not Pinning Teletraan), but might be running.",
```
Comment (yongwen) [inline deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java:171]: pinning -> pinging

**Review by yongwen — APPROVED** [review]

**Pull request merged**